oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/OptimizerEndPoint.kt

```diff
@@ -8,198 +8,215 @@ import kotlinx.coroutines.channels.Channel
+    }
 
     suspend fun startOptimization(optimizer: RandomNumberOptimizer) {
         state = State.Running
-
-        for (simName in modelService.simByName.keys) {
-            syncConfigFor(simName)
-        }
-
         modelService.updateStatusMessage("Evaluating...")
-
         while (state == State.Running) {
-            var toolNumber = 1
-            for ((simName, sim) in modelService.simByName) {
-                modelService.updateStatusMessage("Evaluating: $simName ($toolNumber/${modelService.simByName.keys.size})")
-                val inputVector = optimizer.generateInputs(sim.inputs)
-                val message = OASISQueryDTO.newBuilder()
-                        .setEvaluationRequest(OASISQueryDTO.SimulationEvaluationRequest.newBuilder()
-                                .setName(simName)
-                                .putAllInputVector(inputVector)
-                        )
-                        .build()
-
-                sim.input.onNext(message)
-                val result = select<SimResult> {
-                    sim.output.onReceive { SimResult.Success(it.name, it.outputVectorMap) }
-                    sim.error.onReceive { SimResult.Failure(it.name, it.exception) }
-                    if (sim.timeOut != null) {
-                        onTimeout(sim.timeOut.toMillis()) {
-                            //Timed out
-                            val message = OASISQueryDTO.newBuilder().setCancelRequest(
-                                    OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(simName)
-                            ).build()
-                            sim.input.onNext(message)
-                            select {
-                                sim.output.onReceive { SimResult.TimeOut(it.name) }
-                                sim.error.onReceive { SimResult.TimeOutFailure(it.name, it.exception) }
-                            }
-                        }
-                    }
+            var pluginNumber = 1
+            for ((name, proxy) in modelService.proxyByName) {
+                modelService.updateStatusMessage("Evaluating: $name ($pluginNumber/${modelService.simByName.keys.size})")
+                val inputVector = optimizer.generateInputs(proxy.inputs)
+                val simResult = evaluate(name, inputVector)
+                modelService.addResult(makeResult(simResult, inputVector))
+                modelService.updateStatusMessage("Evaluation finished.")
+                if (simResult is EvaluationResult.TimeOut) {
+                    modelService.updateStatusMessage("Timed out, Canceling...")
+                    cancel(name)
+                    modelService.updateStatusMessage("Cancel finished.")
                 }
+                pluginNumber++
+            }
+            if(state == State.Pause){
+                modelService.updateStatusMessage("Paused")
+                while (state == State.Pause && state != State.Idle){
```

Comment:
Hmm, so do you think this belongs here?

It seems to me like business logic that we would rather put on the `DataModelService` (you can see this by the "feature envy" this class has for `modelService`), but the problem with that is that the two classes would then have a cyclic dependency.

How do you think we should model this? Do you like this current solution?

Reply:
Yikes, this is busy waiting ❗️ 

busy waiting is not ok code to write. 
- From a coroutine context you can use `delay` (in C# its `Task.delay`)
- From _any_ context you can use `Thread.sleep()` (same in C#)

Reply:
regarding the `state` field,

because it can be changed both by the volition API (the netty thread pool) and the UI (the UI thread) this code is fundamentally broken.

There are a couple of solutions we can employ:

1. use coroutines `withContext`
2. use a lock
3. use a CAS loop.

I think it would be very instructional for you to learn about these devices. Let me know when you think thats a good idea!

Reply:
For now, this is going to use a delay